sea-query-derive/src/lib.rs

```diff
@@ -121,126 +236,139 @@ where
 
-    let DeriveInput {
-        ident, data, attrs, ..
-    } = parse_macro_input!(input);
+const DEFAULT_PREFIX: &str = "";
+const DEFAULT_SUFFIX: &str = "Iden";
+const DEFAULT_CRATE_NAME: &str = "sea_query";
+
+impl Default for GenEnumArgs {
+    fn default() -> Self {
+        Self {
+            prefix: Some(DEFAULT_PREFIX.to_string()),
+            suffix: Some(DEFAULT_SUFFIX.to_string()),
+            crate_name: Some(DEFAULT_CRATE_NAME.to_string()),
+            table_name: None,
+        }
+    }
+}
 
-    let table_name = match get_table_name(&ident, attrs) {
+#[proc_macro_attribute]
+pub fn enum_def(args: TokenStream, input: TokenStream) -> TokenStream {
+    let attr_args = match darling::ast::NestedMeta::parse_meta_list(args.into()) {
         Ok(v) => v,
-        Err(e) => return e.to_compile_error().into(),
+        Err(e) => {
+            return TokenStream::from(darling::Error::from(e).write_errors());
+        }
     };
+    let input = parse_macro_input!(input as DeriveInput);
 
-    // Currently we only support enums and unit structs
-    let variants =
-        match data {
-            syn::Data::Enum(DataEnum { variants, .. }) => variants,
-            syn::Data::Struct(DataStruct {
-                fields: Fields::Unit,
-                ..
-            }) => {
-                let impl_iden = impl_iden_for_unit_struct(&ident, &table_name);
-
-                return quote! {
-                    #impl_iden
-
-                    impl #sea_query_path::IdenStatic for #ident {
-                        fn as_str(&self) -> &'static str {
-                            #table_name
-                        }
-                    }
-
-                    impl std::convert::AsRef<str> for #ident {
-                        fn as_ref(&self) -> &str {
-                            self.as_str()
-                        }
-                    }
-                }
-                .into();
-            }
-            _ => return quote_spanned! {
-                ident.span() => compile_error!("you can only derive Iden on enums or unit structs");
-            }
-            .into(),
-        };
-
-    if variants.is_empty() {
-        return TokenStream::new();
-    }
-
-    let impl_iden = impl_iden_for_enum(&ident, &table_name, variants.iter());
+    let args = match GenEnumArgs::from_list(&attr_args) {
+        Ok(v) => v,
+        Err(e) => {
+            return TokenStream::from(e.write_errors());
+        }
+    };
 
-    let match_arms = match variants
-        .iter()
-        .map(|v| (table_name.as_str(), v))
-        .map(IdenVariant::<DeriveIdenStatic>::try_from)
-        .collect::<syn::Result<Vec<_>>>()
-    {
-        Ok(v) => quote! { #(#v),* },
-        Err(e) => return e.to_compile_error().into(),
+    let fields = match &input.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(fields),
+            ..
+        }) => &fields.named,
+        _ => panic!("#[enum_def] can only be used on structs"),
```

Comment:
This should emit a `compile_error!` instead of panicking in the proc-macro

Reply:
Do you mean something like this:
```rust
            _ => return quote_spanned! {
                input.span() => compile_error!("you can only derive Iden on enums or unit structs");
            }
            .into(),
```

Reply:
Yep! The user feedback is better when emitting compiler errors instead of panicking

Reply:
changed in my last commit